heft-plugins/heft-workbox-plugin/src/WorkboxPlugin.ts

```diff
@@ -0,0 +1,121 @@
+
+import { ITerminal, Import } from '@rushstack/node-core-library';
+import type {
+  HeftConfiguration,
+  HeftSession,
+  ICompileSubstage,
+  IBuildStageContext,
+  IHeftPlugin,
+  ScopedLogger
+} from '@rushstack/heft';
+
+import { configurationFileLoader, IWorkboxConfigurationJson } from './WorkboxConfigLoader.js';
+
+import { copyFile } from 'fs';
+
+const workboxBuild: typeof import('workbox-build') = Import.lazy('workbox-build', require);
+
+const PLUGIN_NAME: string = 'WorkboxPlugin';
+
+interface IWorkboxConfigurationFileCacheEntry {
+  configurationFile: IWorkboxConfigurationJson | undefined;
+}
+
+/**
+ * @internal
+ */
+export class WorkboxPlugin implements IHeftPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+
+  private _workboxConfigurationFileCache: Map<string, IWorkboxConfigurationFileCacheEntry> = new Map<
+    string,
+    IWorkboxConfigurationFileCacheEntry
+  >();
+
+  private _logger!: ScopedLogger;
+
+  public apply(heftSession: HeftSession, heftConfiguration: HeftConfiguration): void {
+    this._logger = heftSession.requestScopedLogger('workbox');
+    heftSession.hooks.build.tap(PLUGIN_NAME, (build: IBuildStageContext) => {
+      build.hooks.compile.tap(PLUGIN_NAME, (compile: ICompileSubstage) => {
+        compile.hooks.afterCompile.tapPromise(PLUGIN_NAME, async () => {
+          await this._runWorkboxAsync(heftSession, heftConfiguration, build.properties.production);
+        });
+        compile.hooks.afterRecompile.tapPromise(PLUGIN_NAME, async () => {
+          await this._runWorkboxAsync(heftSession, heftConfiguration, build.properties.production);
+        });
+      });
+    });
+  }
+
+  private async _ensureConfigFileLoadedAsync(
+    terminal: ITerminal,
+    heftConfiguration: HeftConfiguration
+  ): Promise<IWorkboxConfigurationJson | undefined> {
+    const buildFolder: string = heftConfiguration.buildFolder;
+
+    let workboxConfigurationFileCacheEntry: IWorkboxConfigurationFileCacheEntry | undefined =
+      this._workboxConfigurationFileCache.get(buildFolder);
+
+    if (!workboxConfigurationFileCacheEntry) {
+      workboxConfigurationFileCacheEntry = {
+        configurationFile: await configurationFileLoader().tryLoadConfigurationFileForProjectAsync(
+          terminal,
+          buildFolder,
+          heftConfiguration.rigConfig
+        )
+      };
+
+      this._workboxConfigurationFileCache.set(buildFolder, workboxConfigurationFileCacheEntry);
+    }
+
+    return workboxConfigurationFileCacheEntry.configurationFile;
+  }
+
+  private async _runWorkboxAsync(
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration,
+    isProduction: boolean
+  ): Promise<void> {
+    const workboxConfigurationJson: IWorkboxConfigurationJson | undefined =
+      await this._ensureConfigFileLoadedAsync(this._logger.terminal, heftConfiguration);
+    if (!workboxConfigurationJson) {
+      throw new Error('Unable to find an workbox.json config');
+    }
```

Comment:
```suggestion
    if (!workboxConfigurationJson) {
      this._logger.emitError(new Error('Unable to find an workbox.json config'));
      return;
    }
```

Should this be a no-op instead?

Reply:
Ah good question, I wasn't sure if configuration as a plugin in the heft.json should then cause an error if the plugin can't find its config file, or if it should just no-op. I kinda lean more towards the format since it can be confusing when you think you've set something up but haven't? But I guess if we're logging the error too then its clear?

Reply:
Either way works. Just depends on how the plugin is expected to be consumed.

Reply:
Went for your way, and error in the console is sufficient.